platform/src/main/java/org/hillview/sketches/DyadicHistogramBuckets.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2018 VMware Inc. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.hillview.sketches;
+
+import org.hillview.table.api.IColumn;
+
+/**
+ * DyadicHistogramBuckets ensures that every bucket is a multiple of the leaf granularity.
+ * It is built on data that can be converted to doubles.
+ * The last bucket is right-inclusive.
+ */
+public class DyadicHistogramBuckets implements IHistogramBuckets {
+    private double minValue;
+    private double maxValue;
+    private int numOfBuckets;
+    private final double bucketSize;
+    private final double range;
+
+    private final int granularity;
+    private final int numLeaves;
+
+    private void roundMinMax() {
+        long intMin = (long)Math.floor(this.minValue);
+        long intMax = (long)Math.ceil(this.maxValue);
+
+        long minRem = intMin % this.granularity;
+        long maxRem = intMax % this.granularity;
+
+        intMin = intMin - minRem;
+        if ( maxRem > 0 ) { // Don't accidentally add an extra leaf!
+            intMax = intMax + (this.granularity - maxRem);
+        }
+
+        this.minValue = (double)intMin;
+        this.maxValue = (double)intMax;
+    }
+
+    public DyadicHistogramBuckets(final double minValue, final double maxValue,
+                                  final int numOfBuckets, final int granularity) {
+        this.granularity = granularity;
+
+        if (maxValue < minValue || numOfBuckets <= 0)
+            throw new IllegalArgumentException("Negative range or number of buckets");
+        this.minValue = minValue;
+        this.maxValue = maxValue;
+
+        this.range = this.maxValue - this.minValue;
+        if (this.range <= 0)
+            this.numOfBuckets = 1;  // ignore specified number of buckets
+        else
+            this.numOfBuckets = numOfBuckets;
+
+        // Ensure that min and max are an integer multiple of granularity
+        this.roundMinMax();
+        this.bucketSize = (this.maxValue - this.minValue) / this.numOfBuckets;
+        this.numLeaves = (int)((this.maxValue - this.minValue) / this.granularity);
+
+        // Preserves semantics, will make noise computation easier
+        if ( this.numLeaves < this.numOfBuckets ) {
+            this.numOfBuckets = this.numLeaves;
+        }
+    }
+
+    public int indexOf(double value) {
```

Comment:
This code is very similar to DoubleHistogramBuckets.
Can you inherit from that class?

Reply:
Actually, I think that probably this is not how this should be done; it is simpler to first round each value in the dataset down to the closest discretization boundary and then count it. @parikg can explain what I mean if this is unclear.

Reply:
Hmm, these functions will certainly incur floating-point errors if we want to allow non-integer granularity/range, which is dangerous if it leads us to add the wrong amount of noise to a bucket. It's not clear to me how to implement `numLeavesInBucket` safely if `indexOf` is implemented as you suggested. One solution could be to switch to BigDecimal, but I don't think that entirely saves us from precision errors. Let me know if you have other suggestions.

Relatedly, we'll eventually have to implement floating-point-secure Laplace sampling: https://crysp.uwaterloo.ca/courses/pet/F18/cache/Mironov.pdf